### PR TITLE
fix(skore): Support `data_source="both"` for `CrossValidationReport` in `PredictionErrorDisplay`

### LINKED ISSUE #2925: Fix `data_source="both"` in metric displays for cv reports
Follow up to #2549 which was only for the `MetricsSummaryDisplay`.

These are the displays to update: 
- [x] RocCurveDisplay #3006
- [x] PrecisionRecallCurveDisplay #3006
- [ ] PredictionErrorDisplay

### PR DESCRIPTION
**Pull request by direkkakkar319-ops**
#### Change description
Fixes #2925 (PART 2)

We used the `_reoder_categorical_by_appeareance` that we made in https://github.com/probabl-ai/skore/pull/3006 while working towards ROC and PR displayss.

##### what was fixed in this pr 
1. API gap: the `PredictionErrorDisplay` was able to plot `data_source="both"` but was not done officially 
2. Broken Legend: (caused due to https://github.com/probabl-ai/skore/pull/3154/changes#diff-9743801367e9202a4e11d880a61d0dab744a6c2c1262259c8b233dc7d525bd08L330) 
The old code prefixed every label with "Split #", producing junk like:
`Split #split, Split #0, Split #1, Split #data_source, Split #train, Split #test`
insted of 
`Split #0, Split #1, train, test, Perfect predictions`
now only real hue values will be converted as split wise column header , derops the other section headers that not not required by https://github.com/probabl-ai/skore/pull/3154/changes#diff-9743801367e9202a4e11d880a61d0dab744a6c2c1262259c8b233dc7d525bd08R418

##### code block for bug report and fixing it 
```python
import matplotlib
matplotlib.use('Agg')
from sklearn.datasets import load_diabetes
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
from skore import compare, evaluate

X, y = load_diabetes(return_X_y=True)
report = compare([
    evaluate(HistGradientBoostingRegressor(), X, y, splitter=2),
    evaluate(DecisionTreeRegressor(max_depth=3), X, y, splitter=2),
])

display = report.metrics.prediction_error(data_source="both")
fig = display.plot(kind="actual_vs_predicted")
fig.savefig(r'F:\skore_05\prediction_error_with_bug.png', dpi=80, bbox_inches='tight')

```
##### currently 
image

##### after fix
 (change in the order of the graph and legend is also updated)
image

#### Contribution checklist
- [x] Unit tests were added or updated
- [ ] Documentation was added or updated
- [ ] The code passes our style conventions 
- [x] All the tests pass
- [ ] The documentation builds and renders properly
- [x] All the commits in the PR are signed
- [ ] The pull request title respects the Conventional Commits convention

#### AI usage disclosure
AI tools were involved for:
- [ ] Code generation (e.g., when writing an implementation or fixing a bug)
- [x] Test/benchmark generation
- [ ] Documentation (including examples)
- [x] Research and understanding

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py [hunk 1]
```diff
@@ -411,6 +414,34 @@ def _get_plot_columns(
...
+    @staticmethod
+    def _format_relplot_legend(
+        handles,
+        labels: list[str],
+        *,
+        hue: str | None,
+        style: str | None,
+        plot_data: DataFrame,
+    ) -> tuple[list, list[str]]:
+        """Filter seaborn legend entries and format hue labels."""
```
Comment (direkkakkar319-ops) [inline skore/src/skore/_sklearn/_plot/metrics/prediction_error.py:429]: is this one liner of docstring enough

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py [hunk 2]
```diff
@@ -258,8 +259,9 @@ def _plot_matplotlib(
             ]
             y_line = [0, 0]
 
-        plot_data = self.frame()
+        plot_data = self.frame().copy() # Ensure a fresh DataFrame since `frame()` returns a slice
```
Comment (direkkakkar319-ops) [inline skore/src/skore/_sklearn/_plot/metrics/prediction_error.py]: hi @GaetandeCast 
`PredictionErrorDisplay.frame()` returns a column slice view (`self._prediction_error[columns]`). thats why call `.copy()` here specifically in `PredictionErrorDisplay._plot_matplotlib` so that `_reorder_categoricals_by_appearance` does not mutate `self._prediction_error` in place or trigger Pandas `SettingWithCopyWarning`, aslo without unnecessary copying for ROC/PR displays as requested(https://github.com/probabl-ai/skore/pull/3006#discussion_r3413519382)

File: skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py [hunk 3]
```diff
@@ -1152,19 +1152,20 @@ def precision_recall(
...
 
         Parameters
         ----------
-        data_source : {"test", "train"}, default="test"
+        data_source : {"test", "train", "both"}, default="test"
             The data source to use.
 
             - "test" : use the test set provided when creating the report.
             - "train" : use the train set provided when creating the report.
+            - "both" : use both train and test and display them side by side.
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py]:
```suggestion
            - "both" : use both train and test and display them together.
```
Actually we only display them side-by-side when `subplot_by="data_source"`.
Can you apply the same change to the other displays too ?

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py (hunk 2 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/prediction_error.py]: okay 👍

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py (hunk 1 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/prediction_error.py:429]: Yes for a helper such as this one it's fine.

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py [hunk 4]
```diff
@@ -411,6 +414,34 @@ def _get_plot_columns(
 
         return col, hue, style
 
+    @staticmethod
+    def _format_relplot_legend(
+        handles,
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/prediction_error.py]: Please add a type annotation

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py [hunk 5]
```diff
@@ -324,13 +326,14 @@ def _plot_matplotlib(
         handles = []
         labels = []
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/prediction_error.py]: You're gonna have to add type annotations here to please mypy.

File: skore/src/skore/_sklearn/_plot/metrics/prediction_error.py [hunk 6]
```diff
@@ -411,6 +414,34 @@ def _get_plot_columns(
...
+        )
+        formatted_handles = []
+        formatted_labels = []
+        for handle, label in zip(handles, labels, strict=True):
+            if (hue is not None and label == hue) or (style is not None and label == style):
+                continue
+            if hue == "split" and label in hue_values:
+                label = f"Split #{label}"
+            elif hue == "output" and label in hue_values:
+                label = f"Output #{label}"
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/prediction_error.py:445]: Writing this here because this made me try it:
For multi-output regression, we can't distinguish the different splits. For roc and PR it is solved by having multiple curves with the same style but this does not apply here. I can't find a good solution other than adding rows to the plot, which I don't really want us to do. So I think for now we'll keep it like this but I wanted to write it down somewhere.

**Review by GaetandeCast — CHANGES_REQUESTED** [review]
A few remarks then let's merge! Also you have linting with ruff and mypy to fix.

File: skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py (hunk 3 above)
Comment (direkkakkar319-ops) [reply skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py]: applied them to the other reports for the corssvalidation also

**Review by GaetandeCast — APPROVED** [review]
Thanks @direkkakkar319-ops !

**Comment by direkkakkar319-ops** [comment]
Love contributing :) @GaetandeCast
